From 1b09cc193db88afeaa3b58e7b8f3e915ad174da1 Mon Sep 17 00:00:00 2001
From: jayesh <jayesh.babu@windriver,com>
Date: Tue, 8 Mar 2016 21:39:18 +0900
Subject: [PATCH 09/29] P12_0010-clipper_dts_rebase


diff --git a/arch/arm/boot/dts/mb86s72.dtsi b/arch/arm/boot/dts/mb86s72.dtsi
index 332f7d2..cb09a6f 100644
--- a/arch/arm/boot/dts/mb86s72.dtsi
+++ b/arch/arm/boot/dts/mb86s72.dtsi
@@ -1161,6 +1161,7 @@
 	fdb0: fdb.0 {
 		compatible = "fujitsu,f_fdb",
 			     "simple-bus";
+		status = "disabled";
 
 		id = <0>;
 		#stream-id-cells = <2>;
@@ -1224,6 +1225,7 @@
 
 	sdhci0: emmc@300c0000 {
 		compatible = "fujitsu,f_sdh30";
+		status = "disabled";
 		reg = <0 0x300c0000 0x1000>;
 		interrupts = <0 164 0x4>,
 			     <0 165 0x4>;
@@ -1252,6 +1254,7 @@
 
 	sdhci2: sdio@31410000 {
 		compatible = "fujitsu,f_sdh30";
+		status = "disabled";
 		reg = <0 0x31410000 0x1000>;
 		interrupts = <0 176 0x4>,
 			     <0 177 0x4>;
-- 
1.7.1

